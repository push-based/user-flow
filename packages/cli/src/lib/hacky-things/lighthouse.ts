/*
* NOTICE:
*
* As long as we did bot solve the import problem we will isolate this hack here to not pollute our codebase with `@ts-ignore`.
*
* */

// @ts-ignore
export {Util} from 'lighthouse/lighthouse-core/util-commonjs';
// @ts-ignore
export { startFlow, UserFlow } from 'lighthouse/lighthouse-core/fraggle-rock/api';
