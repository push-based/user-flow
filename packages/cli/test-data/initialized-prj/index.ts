export * from './cfg';
export * from './user-flow.uf';
