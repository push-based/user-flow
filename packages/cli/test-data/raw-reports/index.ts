import { join } from "path";
import { readFileSync } from "fs";

export function getReportContent(fileName: string): string | {} {
  const path = join(__dirname, fileName);
  const report = readFileSync(path, 'utf-8');
  if(fileName.endsWith('.json')) {
    JSON.parse(report);
  }
  return report;
}
