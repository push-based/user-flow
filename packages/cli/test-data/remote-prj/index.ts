export * from './cfg';
export * from './rc.json';
export * from './user-flow.uf';
