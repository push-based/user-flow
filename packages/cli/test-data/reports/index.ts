import { readFileSync } from "fs";
import { join } from "path";

export const LHR9_HTML_NAME = 'lhr-9.html';
export const LHR9_JSON_NAME = 'lhr-9.json';


export function getContent(fileName: string): string | {} {
  const reportContent = readFileSync(join(__dirname, fileName), 'utf-8');
  return reportContent;
}
