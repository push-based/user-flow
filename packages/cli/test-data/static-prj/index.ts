export * from './dist';
export * from './cfg';
export * from './rc.json';
export * from './user-flow.uf';
