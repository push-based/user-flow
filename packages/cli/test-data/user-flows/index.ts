export * from './valide.example.uf';
export * from './wrong-ext.example.uf';
export * from './wrong-mod-export.example.uf';
