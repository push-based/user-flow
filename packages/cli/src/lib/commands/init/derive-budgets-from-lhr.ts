import { UserFlow } from '../../hacky-things/lighthouse.js';
// @ts-ignore
import Budget from 'lighthouse/types/lhr/budget';
import { RequestCountResourceTypeBudgets, TransferSizeResourceTypeBudgets } from './constants.js';
import { logVerbose } from '../../core/loggin/index.js';

export function deriveBudgetsFromLhr(flow: any): Budget[] {
  const budgetObject: Budget = {};
  if (flow?.steps[0].lhr?.audits) {
    if (flow.steps[0].lhr.audits['resource-summary']) {
      const resourceSummary = flow.steps[0].lhr.audits['resource-summary'];
      budgetObject.resourceSizes = (resourceSummary.details.items as any)
        .filter(({ resourceType }: any) => TransferSizeResourceTypeBudgets.includes(resourceType))
        .map(({ resourceType, transferSize }: any) => ({
            resourceType,
            budget: transferSize
          })
        );
      budgetObject.resourceCounts = (resourceSummary.details.items as any)
        .filter(({ resourceType }: any) => RequestCountResourceTypeBudgets.includes(resourceType))
        .map(({ resourceType, requestCount }: any) => ({
            resourceType,
            budget: requestCount
          })
        );
    }
    else {
      logVerbose(`The lighthouse report does not contain a "resource-summary" audit`);
    }
    budgetObject.timings = [];
    if (flow.steps[0].lhr.audits['cumulative-layout-shift']) {
      budgetObject.timings.push({
        'metric': 'cumulative-layout-shift',
        'budget': flow.steps[0].lhr.audits['cumulative-layout-shift'].details.items[0].totalCumulativeLayoutShift
      });
    } else {
      logVerbose(`The lighthouse report does not contain a "cumulative-layout-shift" audit`);
    }

    if (flow.steps[0].lhr.audits['largest-contentful-paint']) {
      budgetObject.timings.push({
        'metric': 'largest-contentful-paint',
        'budget': parseInt(flow.steps[0].lhr.audits['largest-contentful-paint'].numericValue)
      });
    } else {
      logVerbose(`The lighthouse report does not contain a "largest-contentful-paint" audit`);
    }

  }
  return [budgetObject];
}

