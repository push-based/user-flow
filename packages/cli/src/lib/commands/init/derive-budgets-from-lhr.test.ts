import { describe, expect, it } from 'vitest';
import { existsSync } from 'fs';
import { join } from 'path';
import { deriveBudgetsFromLhr } from './derive-budgets-from-lhr.js';
import { getReportContent } from 'test-data';
import { FlowResult } from 'lighthouse';

const lhr9 = getReportContent<FlowResult>('lhr-9.json');
const expectedBudgets = [
  {
    resourceCounts: [
      { budget: 0, resourceType: 'stylesheet' },
      { budget: 0, resourceType: 'script' }
    ],
    resourceSizes: [
      { budget: 941, resourceType: 'document' },
      { budget: 0, resourceType: 'stylesheet' },
      { budget: 0, resourceType: 'font' },
      { budget: 0, resourceType: 'script' },
      { budget: 0, resourceType: 'third-party' }
    ],
    timings: [
      { budget: 0, metric: 'cumulative-layout-shift' },
      { budget: 724, metric: 'largest-contentful-paint' }
    ]
  }
];

const expectedFilePath = join('.github', 'workflows', 'user-flow-ci.yml');
describe('deriveBudgetsFromLhr', () => {

  it('should create budgets array if a valid lhr is given', async () => {
    expect(existsSync(expectedFilePath)).toBeFalsy();
    const budgets = deriveBudgetsFromLhr(lhr9);
    expect(budgets).toStrictEqual(expectedBudgets);
  });

});
