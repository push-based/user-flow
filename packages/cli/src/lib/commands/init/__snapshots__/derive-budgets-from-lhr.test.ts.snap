// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`deriveBudgetsFromLhr > should create budgets array if a valid lhr is given 1`] = `
[
  {
    "resourceCounts": [
      {
        "budget": 0,
        "resourceType": "stylesheet",
      },
      {
        "budget": 0,
        "resourceType": "script",
      },
    ],
    "resourceSizes": [
      {
        "budget": 941,
        "resourceType": "document",
      },
      {
        "budget": 0,
        "resourceType": "stylesheet",
      },
      {
        "budget": 0,
        "resourceType": "font",
      },
      {
        "budget": 0,
        "resourceType": "script",
      },
      {
        "budget": 0,
        "resourceType": "third-party",
      },
    ],
    "timings": [
      {
        "budget": 0,
        "metric": "cumulative-layout-shift",
      },
      {
        "budget": 724,
        "metric": "largest-contentful-paint",
      },
    ],
  },
]
`;
